<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Para-guardar-o-dataset-numa-BD-SQL,-utilizamos-a-library-sqlite3,-e-executamos-os-seguintes-códigos:" data-toc-modified-id="Para-guardar-o-dataset-numa-BD-SQL,-utilizamos-a-library-sqlite3,-e-executamos-os-seguintes-códigos:-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Para guardar o dataset numa BD SQL, utilizamos a library sqlite3, e executamos os seguintes códigos:</a></span></li><li><span><a href="#Posteriormente,-para-selecionar-apenas-os-200-primeiros-registos-da-base-de-dados-préviamente-criada-e-já-com-a-coneção-efetuada:" data-toc-modified-id="Posteriormente,-para-selecionar-apenas-os-200-primeiros-registos-da-base-de-dados-préviamente-criada-e-já-com-a-coneção-efetuada:-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Posteriormente, para selecionar apenas os 200 primeiros registos da base de dados préviamente criada e já com a coneção efetuada:</a></span></li><li><span><a href="#Para-selecionar-todos-os-registos-dos-campos-id,-year,-month,-element,-value1,-value2,-value3,-value4,-value5,-value6,-value7,-value8,-value9,-value10,-value11,-value12,-value13,-value14,-value15,-value16,-value17,-value18,-value19,-value20,-value21,-value22,-value23,-value24,-value25,-value26,-value27,-value28,-value29,-value30,-value31-das-cinco-estações-meteorológicas-portuguesas-(Horta;-Funchal;-Lisboa;-Castelo-Branco-e-Faro)-que-existem-no-dataset,-corremos-a-seguinte-query:" data-toc-modified-id="Para-selecionar-todos-os-registos-dos-campos-id,-year,-month,-element,-value1,-value2,-value3,-value4,-value5,-value6,-value7,-value8,-value9,-value10,-value11,-value12,-value13,-value14,-value15,-value16,-value17,-value18,-value19,-value20,-value21,-value22,-value23,-value24,-value25,-value26,-value27,-value28,-value29,-value30,-value31-das-cinco-estações-meteorológicas-portuguesas-(Horta;-Funchal;-Lisboa;-Castelo-Branco-e-Faro)-que-existem-no-dataset,-corremos-a-seguinte-query:-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Para selecionar todos os registos dos campos id, year, month, element, value1, value2, value3, value4, value5, value6, value7, value8, value9, value10, value11, value12, value13, value14, value15, value16, value17, value18, value19, value20, value21, value22, value23, value24, value25, value26, value27, value28, value29, value30, value31 das cinco estações meteorológicas portuguesas (Horta; Funchal; Lisboa; Castelo Branco e Faro) que existem no dataset, corremos a seguinte query:</a></span></li><li><span><a href="#Substitua-os-ids-presentes-no-dataframe-pelo-correspondente-nome-da-estação-meteorológica." data-toc-modified-id="Substitua-os-ids-presentes-no-dataframe-pelo-correspondente-nome-da-estação-meteorológica.-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Substitua os ids presentes no dataframe pelo correspondente nome da estação meteorológica.</a></span></li><li><span><a href="#Aplique-as-funções-de-data-wrangling-e-de-compression-dos-dados-que-achar-adequadas." data-toc-modified-id="Aplique-as-funções-de-data-wrangling-e-de-compression-dos-dados-que-achar-adequadas.-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Aplique as funções de data wrangling e de compression dos dados que achar adequadas.</a></span></li><li><span><a href="#Determine-a-percentagem-de-null-values-em-cada-uma-das-variáveis." data-toc-modified-id="Determine-a-percentagem-de-null-values-em-cada-uma-das-variáveis.-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Determine a percentagem de null values em cada uma das variáveis.</a></span></li><li><span><a href="#Crie-uma-nova-coluna-denominada-count_nan-que-registe-para-cada-observação-o-número-de-dias-com-temperaturas-máximas-nulas-(NaN)." data-toc-modified-id="Crie-uma-nova-coluna-denominada-count_nan-que-registe-para-cada-observação-o-número-de-dias-com-temperaturas-máximas-nulas-(NaN).-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Crie uma nova coluna denominada count_nan que registe para cada observação o número de dias com temperaturas máximas nulas (NaN).</a></span></li><li><span><a href="#Crie-um-novo-dataframe-que-considere-apenas-os-registos-com-um-valor-total-de-NaN-inferior-a-7." data-toc-modified-id="Crie-um-novo-dataframe-que-considere-apenas-os-registos-com-um-valor-total-de-NaN-inferior-a-7.-8"><span class="toc-item-num">8&nbsp;&nbsp;</span>Crie um novo dataframe que considere apenas os registos com um valor total de NaN inferior a 7.</a></span></li><li><span><a href="#Reccorra-agora-a-chunking-para-determinar-a-percentagem-de-null-values-em-cada-uma-das-variáveis-na-totalidade-dos-registos-do-dataframe." data-toc-modified-id="Reccorra-agora-a-chunking-para-determinar-a-percentagem-de-null-values-em-cada-uma-das-variáveis-na-totalidade-dos-registos-do-dataframe.-9"><span class="toc-item-num">9&nbsp;&nbsp;</span>Reccorra agora a chunking para determinar a percentagem de null values em cada uma das variáveis na totalidade dos registos do dataframe.</a></span></li><li><span><a href="#Determine-a-temperatura-média-para-cada-observação-(estação/ano/mês)-do-dataframe.-Guarde-os-valores-obtidos-numa-nova-coluna-denominada-daily_avg_temp.-Não-se-esqueça-de-considerar-apenas-as-colunas-cujo-nome-começa-por-value." data-toc-modified-id="Determine-a-temperatura-média-para-cada-observação-(estação/ano/mês)-do-dataframe.-Guarde-os-valores-obtidos-numa-nova-coluna-denominada-daily_avg_temp.-Não-se-esqueça-de-considerar-apenas-as-colunas-cujo-nome-começa-por-value.-10"><span class="toc-item-num">10&nbsp;&nbsp;</span>Determine a temperatura média para cada observação (estação/ano/mês) do dataframe. Guarde os valores obtidos numa nova coluna denominada daily_avg_temp. Não se esqueça de considerar apenas as colunas cujo nome começa por value.</a></span></li><li><span><a href="#Agrupe-os-dados-por-nome-de-estação-e-ano-aplicando-a-função-mean.-Mostre-apenas-os-resultados-para-a-coluna-daily_avg_temp." data-toc-modified-id="Agrupe-os-dados-por-nome-de-estação-e-ano-aplicando-a-função-mean.-Mostre-apenas-os-resultados-para-a-coluna-daily_avg_temp.-11"><span class="toc-item-num">11&nbsp;&nbsp;</span>Agrupe os dados por nome de estação e ano aplicando a função mean. Mostre apenas os resultados para a coluna daily_avg_temp.</a></span></li></ul></div>

In [122]:
# Importar Módulos
import pandas as pd
import sqlite3 as sql
import os
import numpy as np

In [37]:
# Caminho e nome do ficheiro
current_dir = os.getcwd()
ignored_dir = os.path.join(current_dir, 'ignored_folder')
ghcnd_daily_csv_path = os.path.join(ignored_dir, 'global_climate_data', 'ghcnd_daily.csv')

print(ghcnd_daily_csv_path)

/mnt/storage/Academic/IPT/2022-2023 - 1st Semestre/Big Data/Project/Big_data_IPT/ignored_folder/global_climate_data/ghcnd_daily.csv


In [38]:
#
if os.path.exists(ghcnd_daily_csv_path):
    chunk_size=50000
    chunk = pd.read_csv(ghcnd_daily_csv_path, chunksize=chunk_size, low_memory=False)
    print(type(chunk))

<class 'pandas.io.parsers.readers.TextFileReader'>


In [39]:
#
df = chunk.get_chunk(chunk_size)
df.head()

id  year  month element  value1  mflag1 qflag1 sflag1  value2  \
0  ACW00011604  1949      1    TMAX     289     NaN    NaN      X     289   
1  ACW00011604  1949      2    TMAX     267     NaN    NaN      X     278   
2  ACW00011604  1949      3    TMAX     272     NaN    NaN      X     289   
3  ACW00011604  1949      4    TMAX     278     NaN    NaN      X     283   
4  ACW00011604  1949      5    TMAX     283     NaN    NaN      X     283   

   mflag2  ... qflag29 sflag29  value30  mflag30 qflag30 sflag30  value31  \
0     NaN  ...     NaN       X      272      NaN     NaN       X      272   
1     NaN  ...     NaN     NaN    -9999      NaN     NaN     NaN    -9999   
2     NaN  ...     NaN       X      278      NaN     NaN       X      267   
3     NaN  ...     NaN       X      289      NaN     NaN       X    -9999   
4     NaN  ...     NaN       X      294      NaN     NaN       X      300   

   mflag31 qflag31 sflag31  
0      NaN     NaN       X  
1      NaN     NaN     NaN  
2      NaN     NaN       X  
3      NaN     NaN     NaN  
4      NaN     NaN       X  

[5 rows x 128 columns]

In [40]:
#
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Columns: 128 entries, id to sflag31
dtypes: float64(31), int64(33), object(64)
memory usage: 48.8+ MB


### Para guardar o dataset numa BD SQL, utilizamos a library sqlite3, e executamos os seguintes códigos:

In [84]:
# Conexão ao ficheiro da BD
ghcnd_daily_db_path = os.path.join(ignored_dir, "ghcnd_daily.db")
conn = sql.connect(ghcnd_daily_db_path)

In [53]:
run = False # change to True to run

if run:
    # Dividir em ficheiros mais pequenos, neste caso, com 300000 linhas
    chunk_size = 300000
    batch_no = 1
    for chunk in pd.read_csv(ghcnd_daily_csv_path, chunksize=chunk_size, iterator=True, low_memory=False):
        chunk.to_sql('ghcnd', conn, if_exists='append')
        print(f'processed batch # {batch_no}')
        batch_no += 1

index: 2
index: 3
index: 4
index: 5
index: 6
index: 7
index: 8
index: 9
index: 10
index: 11
index: 12
index: 13
index: 14
index: 15
index: 16
index: 17
index: 18
index: 19
index: 20
index: 21
index: 22
index: 23
index: 24
index: 25
index: 26
index: 27
index: 28
index: 29
index: 30
index: 31
index: 32
index: 33
index: 34
index: 35
index: 36
index: 37
index: 38
index: 39
index: 40
index: 41
index: 42
index: 43
index: 44
index: 45
index: 46
index: 47
index: 48
index: 49
index: 50
index: 51


Houve tentativa de fazer com 500000, mas ficou muito lento a processar.

### Posteriormente, para selecionar apenas os 200 primeiros registos da base de dados préviamente criada e já com a coneção efetuada:

In [85]:
#
ghcnd = pd.read_sql('SELECT * FROM ghcnd LIMIT 200', conn)
ghcnd.head()

index           id  year  month element  value1 mflag1 qflag1 sflag1  \
0      0  ACW00011604  1949      1    TMAX     289   None   None      X   
1      1  ACW00011604  1949      2    TMAX     267   None   None      X   
2      2  ACW00011604  1949      3    TMAX     272   None   None      X   
3      3  ACW00011604  1949      4    TMAX     278   None   None      X   
4      4  ACW00011604  1949      5    TMAX     283   None   None      X   

   value2 mflag2 qflag2 sflag2  value3 mflag3 qflag3 sflag3  value4 mflag4  \
0     289   None   None      X     283   None   None      X     283   None   
1     278   None   None      X     272   None   None      X     267   None   
2     289   None   None      X     278   None   None      X     278   None   
3     283   None   None      X     283   None   None      X     294   None   
4     283   None   None      X     278   None   None      X     272   None   

  qflag4 sflag4  value5 mflag5 qflag5 sflag5  value6 mflag6 qflag6 sflag6  \
0   None      X     289   None   None      X     289   None   None      X   
1   None      X     278   None   None      X     278   None   None      X   
2   None      X     278   None   None      X     278   None   None      X   
3   None      X     283   None   None      X     283   None   None      X   
4   None      X     278   None   None      X     283   None   None      X   

   value7 mflag7 qflag7 sflag7  value8 mflag8 qflag8 sflag8  value9 mflag9  \
0     278   None   None      X     267   None   None      X     272   None   
1     278   None   None      X     267   None   None      X     278   None   
2     289   None   None      X     294   None   None      X     294   None   
3     278   None   None      X     283   None   None      X     283   None   
4     278   None   None      X     294   None   None      X     294   None   

  qflag9 sflag9  value10 mflag10 qflag10 sflag10  value11 mflag11 qflag11  \
0   None      X      278    None    None       X      267    None    None   
1   None      X      267    None    None       X      272    None    None   
2   None      X      272    None    None       X      272    None    None   
3   None      X      289    None    None       X      283    None    None   
4   None      X      294    None    None       X      289    None    None   

  sflag11  value12 mflag12 qflag12 sflag12  value13 mflag13 qflag13 sflag13  \
0       X      278    None    None       X      267    None    None       X   
1       X      267    None    None       X      261    None    None       X   
2       X      278    None    None       X      283    None    None       X   
3       X      289    None    None       X      289    None    None       X   
4       X      294    None    None       X      294    None    None       X   

   value14 mflag14 qflag14 sflag14  value15 mflag15 qflag15 sflag15  value16  \
0      267    None    None       X      278    None    None       X      267   
1      261    None    None       X      261    None    None       X      267   
2      278    None    None       X      283    None    None       X      283   
3      289    None    None       X      294    None    None       X      289   
4      294    None    None       X      289    None    None       X      294   

  mflag16 qflag16 sflag16  value17 mflag17 qflag17 sflag17  value18 mflag18  \
0    None    None       X      267    None    None       X      272    None   
1    None    None       X      278    None    None       X      272    None   
2    None    None       X      283    None    None       X      289    None   
3    None    None       X      289    None    None       X      294    None   
4    None    None       X      294    None    None       X      294    None   

  qflag18 sflag18  value19 mflag19 qflag19 sflag19  value20 mflag20 qflag20  \
0    None       X      272    None    None       X      272    None    None   
1    None       X      278    None    None       X      278    None    None   
2    None       X   

In [86]:
print(type(ghcnd))
print(ghcnd.info)

<class 'pandas.core.frame.DataFrame'>
<bound method DataFrame.info of      index           id  year  month element  value1 mflag1 qflag1 sflag1  \
0        0  ACW00011604  1949      1    TMAX     289   None   None      X   
1        1  ACW00011604  1949      2    TMAX     267   None   None      X   
2        2  ACW00011604  1949      3    TMAX     272   None   None      X   
3        3  ACW00011604  1949      4    TMAX     278   None   None      X   
4        4  ACW00011604  1949      5    TMAX     283   None   None      X   
..     ...          ...   ...    ...     ...     ...    ...    ...    ...   
195    195  AE000041196  1981      9    TMAX   -9999   None   None   None   
196    196  AE000041196  1981     10    TMAX   -9999   None   None   None   
197    197  AE000041196  1981     11    TMAX     330   None   None      I   
198    198  AE000041196  1981     12    TMAX     270   None   None      I   
199    199  AE000041196  1982      1    TMAX     245   None   None      I   

     

### Para selecionar todos os registos dos campos id, year, month, element, value1, value2, value3, value4, value5, value6, value7, value8, value9, value10, value11, value12, value13, value14, value15, value16, value17, value18, value19, value20, value21, value22, value23, value24, value25, value26, value27, value28, value29, value30, value31 das cinco estações meteorológicas portuguesas (Horta; Funchal; Lisboa; Castelo Branco e Faro) que existem no dataset, corremos a seguinte query:

In [105]:
conn = sql.connect(ghcnd_daily_db_path)

#
ghcnd = pd.read_sql('select id, year, month, element, value1, value2, value3, value4, value5, value6, value7, value8, value9, value10,value11, value12, value13, value14, value15, value16, value17, value18, value19, value20, value21, value22, value23, value24, value25, value26, value27, value28, value29, value30, value31 from ghcnd WHERE id IN ("PO000008506","PO000008522","PO000008535","POM00008570","POM00008554")', conn)

#
conn.close()

#
ghcnd

id  year  month element  value1  value2  value3  value4  \
0     PO000008506  1973      1    TMAX   -9999     110   -9999     170   
1     PO000008506  1973      2    TMAX     170     160     120     110   
2     PO000008506  1973      3    TMAX   -9999     150     160   -9999   
3     PO000008506  1973      4    TMAX     180     170   -9999     190   
4     PO000008506  1973      5    TMAX     140   -9999   -9999     150   
...           ...   ...    ...     ...     ...     ...     ...     ...   
2637  POM00008570  2018      8    TMAX   -9999     422   -9999   -9999   
2638  POM00008570  2018      9    TMAX   -9999   -9999   -9999   -9999   
2639  POM00008570  2018     10    TMAX     297     283   -9999   -9999   
2640  POM00008570  2019      3    TMAX   -9999   -9999   -9999   -9999   
2641  POM00008570  2019      7    TMAX   -9999   -9999   -9999   -9999   

      value5  value6  value7  value8  value9  value10  value11  value12  \
0        160     160   -9999   -9999   -9999      160      170      140   
1        130   -9999   -9999     160     150    -9999    -9999    -9999   
2        160   -9999     160     140     130      120    -9999      140   
3        170   -9999     190   -9999     160    -9999      180      180   
4      -9999     190     190   -9999   -9999      190    -9999      180   
...      ...     ...     ...     ...     ...      ...      ...      ...   
2637   -9999   -9999   -9999   -9999   -9999    -9999    -9999    -9999   
2638   -9999   -9999   -9999   -9999   -9999    -9999    -9999    -9999   
2639     311   -9999   -9999   -9999   -9999    -9999    -9999    -9999   
2640   -9999   -9999   -9999   -9999   -9999    -9999    -9999    -9999   
2641     290   -9999   -9999   -9999   -9999    -9999    -9999    -9999   

      value13  value14  value15  value16  value17  value18  value19  value20  \
0       -9999    -9999      180      140    -9999    -9999      150    -9999   
1         150      170    -9999      170      170      180    -9999      150   
2         150      160      170      150      160      160      160      190   
3       -9999      150    -9999      160    -9999      150      180      170   
4       -9999      180      190      190      170      190      170      160   
...       ...      ...      ...      ...      ...      ...      ...      ...   
2637    -9999    -9999    -9999    -9999    -9999    -9999    -9999    -9999   
2638    -9999    -9999      333      334    -9999    -9999    -9999    -9999   
2639    -9999    -9999    -9999    -9999      221    -9999      212    -9999   
2640      157    -9999    -9999    -9999    -9999    -9999    -9999    -9999   
2641    -9999    -9999    -9999    -9999    -9999    -9999    -9999    -9999   

      value21  value22  value23  value24  value25  value26  value27  value28  \
0         170    -9999    -9999    -9999      130    -9999    -9999    -9999   
1       -9999    -9999      150    -9999      170      180      170      170   
2         160    -9999    -9999      160      170      170      180    -9999   
3         170    -9999      160      170    -9999    -9999      170      170   
4       -9999      180    -9999      190    -9999      190      180      190   
...       ...      ...      ...      ...      ...      ...      ...      ...   
2637    -9999    -9999    -9999    -9999    -9999    -9999    -9999    -9999   
2638      327    -9999    -9999    -9999    -9999      308    -9999    -9999   
2639    -9999    -9999    -9999    -9999    -9999    -9999    -9999    -9999   
2640      177    -9999    -9999    -9999    -9999    -9999    -9999    -9999   
2641    -9999    -9999    -9999    -9999    -9999    -9999    -9999    -9999   

      value29  value30  value31  
0       -9999      180    -9999  
1       -9999    -9999    -9999  
2       -9999    -9999      180  
3         180      180    -9999  
4       -9999      210    -9999  
...       ...      ...      ...  
2637    -9999    -9999    -9999  
2638    -9999    -9999    -9999

### Substitua os ids presentes no dataframe pelo correspondente nome da estação meteorológica.

In [59]:
#
ghcnd['id'] = ghcnd['id'].astype(str)
ghcnd.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2642 entries, 0 to 2641
Data columns (total 35 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   id       2642 non-null   object
 1   year     2642 non-null   int64 
 2   month    2642 non-null   int64 
 3   element  2642 non-null   object
 4   value1   2642 non-null   int64 
 5   value2   2642 non-null   int64 
 6   value3   2642 non-null   int64 
 7   value4   2642 non-null   int64 
 8   value5   2642 non-null   int64 
 9   value6   2642 non-null   int64 
 10  value7   2642 non-null   int64 
 11  value8   2642 non-null   int64 
 12  value9   2642 non-null   int64 
 13  value10  2642 non-null   int64 
 14  value11  2642 non-null   int64 
 15  value12  2642 non-null   int64 
 16  value13  2642 non-null   int64 
 17  value14  2642 non-null   int64 
 18  value15  2642 non-null   int64 
 19  value16  2642 non-null   int64 
 20  value17  2642 non-null   int64 
 21  value18  2642 non-null   int64 
 22  

PO000008506  38.5200  -28.6300   62.0    HORTA (AZORES)                 GSN     08506
PO000008522  32.6300  -16.8997   25.0    FUNCHAL                        GSN     08522
PO000008535  38.7167   -9.1500   77.0    LISBOA GEOFISICA               GSN     08535
POM00008570  39.8330   -7.4830  384.0    CASTELO BRANCO                         08570
POM00008554  37.0140   -7.9660    7.3    FARO                                   08554

In [106]:
#
ghcnd['id'] = ghcnd['id'].replace('PO000008506', 'HORTA').replace('PO000008522', 'FUNCHAL').replace('PO000008535', 'LISBOA GEOFISICA').replace('POM00008570', 'CASTELO BRANCO').replace('POM00008554', 'FARO')

ghcnd

id  year  month element  value1  value2  value3  value4  \
0              HORTA  1973      1    TMAX   -9999     110   -9999     170   
1              HORTA  1973      2    TMAX     170     160     120     110   
2              HORTA  1973      3    TMAX   -9999     150     160   -9999   
3              HORTA  1973      4    TMAX     180     170   -9999     190   
4              HORTA  1973      5    TMAX     140   -9999   -9999     150   
...              ...   ...    ...     ...     ...     ...     ...     ...   
2637  CASTELO BRANCO  2018      8    TMAX   -9999     422   -9999   -9999   
2638  CASTELO BRANCO  2018      9    TMAX   -9999   -9999   -9999   -9999   
2639  CASTELO BRANCO  2018     10    TMAX     297     283   -9999   -9999   
2640  CASTELO BRANCO  2019      3    TMAX   -9999   -9999   -9999   -9999   
2641  CASTELO BRANCO  2019      7    TMAX   -9999   -9999   -9999   -9999   

      value5  value6  value7  value8  value9  value10  value11  value12  \
0        160     160   -9999   -9999   -9999      160      170      140   
1        130   -9999   -9999     160     150    -9999    -9999    -9999   
2        160   -9999     160     140     130      120    -9999      140   
3        170   -9999     190   -9999     160    -9999      180      180   
4      -9999     190     190   -9999   -9999      190    -9999      180   
...      ...     ...     ...     ...     ...      ...      ...      ...   
2637   -9999   -9999   -9999   -9999   -9999    -9999    -9999    -9999   
2638   -9999   -9999   -9999   -9999   -9999    -9999    -9999    -9999   
2639     311   -9999   -9999   -9999   -9999    -9999    -9999    -9999   
2640   -9999   -9999   -9999   -9999   -9999    -9999    -9999    -9999   
2641     290   -9999   -9999   -9999   -9999    -9999    -9999    -9999   

      value13  value14  value15  value16  value17  value18  value19  value20  \
0       -9999    -9999      180      140    -9999    -9999      150    -9999   
1         150      170    -9999      170      170      180    -9999      150   
2         150      160      170      150      160      160      160      190   
3       -9999      150    -9999      160    -9999      150      180      170   
4       -9999      180      190      190      170      190      170      160   
...       ...      ...      ...      ...      ...      ...      ...      ...   
2637    -9999    -9999    -9999    -9999    -9999    -9999    -9999    -9999   
2638    -9999    -9999      333      334    -9999    -9999    -9999    -9999   
2639    -9999    -9999    -9999    -9999      221    -9999      212    -9999   
2640      157    -9999    -9999    -9999    -9999    -9999    -9999    -9999   
2641    -9999    -9999    -9999    -9999    -9999    -9999    -9999    -9999   

      value21  value22  value23  value24  value25  value26  value27  value28  \
0         170    -9999    -9999    -9999      130    -9999    -9999    -9999   
1       -9999    -9999      150    -9999      170      180      170      170   
2         160    -9999    -9999      160      170      170      180    -9999   
3         170    -9999      160      170    -9999    -9999      170      170   
4       -9999      180    -9999      190    -9999      190      180      190   
...       ...      ...      ...      ...      ...      ...      ...      ...   
2637    -9999    -9999    -9999    -9999    -9999    -9999    -9999    -9999   
2638      327    -9999    -9999    -9999    -9999      308    -9999    -9999   
2639    -9999    -9999    -9999    -9999    -9999    -9999    -9999    -9999   
2640      177    -9999    -9999    -9999    -9999    -9999    -9999    -9999   
2641    -9999    -9999    -9999    -9999    -9999    -9999    -9999    -9999   

      value29  value30  value31  
0       -9999      180    -9999  
1       -9999    -9999    -9999  
2       -9999    -9999      180  
3         180      180    -9999  
4       -9999      210    -9999  
...       ...      ...      ...  
2637    -9999    -9999    -9999 

### Aplique as funções de data wrangling e de compression dos dados que achar adequadas.

In [107]:
ghcnd.memory_usage()

Index        128
id         21136
year       21136
month      21136
element    21136
value1     21136
value2     21136
value3     21136
value4     21136
value5     21136
value6     21136
value7     21136
value8     21136
value9     21136
value10    21136
value11    21136
value12    21136
value13    21136
value14    21136
value15    21136
value16    21136
value17    21136
value18    21136
value19    21136
value20    21136
value21    21136
value22    21136
value23    21136
value24    21136
value25    21136
value26    21136
value27    21136
value28    21136
value29    21136
value30    21136
value31    21136
dtype: int64

In [90]:
# Para visualizar todas as colunas
pd.set_option('display.max_columns', None)
ghcnd.head(30)

id  year  month element  value1  value2  value3  value4  value5  \
0   HORTA  1973      1    TMAX   -9999     110   -9999     170     160   
1   HORTA  1973      2    TMAX     170     160     120     110     130   
2   HORTA  1973      3    TMAX   -9999     150     160   -9999     160   
3   HORTA  1973      4    TMAX     180     170   -9999     190     170   
4   HORTA  1973      5    TMAX     140   -9999   -9999     150   -9999   
5   HORTA  1973      6    TMAX   -9999   -9999   -9999     200     200   
6   HORTA  1973      7    TMAX     210   -9999     230     220     240   
7   HORTA  1973      8    TMAX     220     230     230     250   -9999   
8   HORTA  1973      9    TMAX   -9999     260     340     230     240   
9   HORTA  1973     10    TMAX   -9999     220     230     210     230   
10  HORTA  1973     11    TMAX   -9999   -9999   -9999     190     200   
11  HORTA  1973     12    TMAX   -9999   -9999     210     200   -9999   
12  HORTA  1974      1    TMAX     160     170     180     150   -9999   
13  HORTA  1974      2    TMAX     130   -9999     160   -9999     180   
14  HORTA  1974      3    TMAX     170     190     170     160     110   
15  HORTA  1974      4    TMAX     130     160   -9999     230     160   
16  HORTA  1974      5    TMAX     190   -9999   -9999   -9999     200   
17  HORTA  1974      6    TMAX     200     220   -9999   -9999   -9999   
18  HORTA  1974      7    TMAX   -9999   -9999     220     220     220   
19  HORTA  1974      8    TMAX     210     240     230     230     240   
20  HORTA  1974      9    TMAX     250   -9999     250     250     270   
21  HORTA  1974     10    TMAX     220     230     200     200   -9999   
22  HORTA  1974     11    TMAX   -9999     180     200     210     210   
23  HORTA  1974     12    TMAX     200     190   -9999   -9999     200   
24  HORTA  1975      1    TMAX   -9999   -9999   -9999   -9999   -9999   
25  HORTA  1975      2    TMAX   -9999   -9999     160     140     160   
26  HORTA  1975      3    TMAX     160     150     150   -9999     170   
27  HORTA  1975      4    TMAX   -9999     300     180     170   -9999   
28  HORTA  1975      5    TMAX     190     180   -9999     200   -9999   
29  HORTA  1975      6    TMAX     210     210   -9999   -9999     200   

    value6  value7  value8  value9  value10  value11  value12  value13  \
0      160   -9999   -9999   -9999      160      170      140    -9999   
1    -9999   -9999     160     150    -9999    -9999    -9999      150   
2    -9999     160     140     130      120    -9999      140      150   
3    -9999     190   -9999     160    -9999      180      180    -9999   
4      190     190   -9999   -9999      190    -9999      180    -9999   
5      190     200     220   -9999      200    -9999      220      200   
6      250     240     250     210      240      220      210      220   
7      240     250     250     250      240    -9999    -9999      260   
8      250   -9999   -9999   -9999      230    -9999      240      220   
9    -9999   -9999     240     230      230    -9999    -9999    -9999   
10   -9999   -9999     210     210    -9999      210    -9999      180   
11     220   -9999   -9999     190    -9999    -9999      190    -9999   
12     170     180   -9999   -9999      170    -9999      140      170   
13   -9999   -9999   -9999     170    -9999    -9999    -9999      180   
14     130     170     160   -9999      160    -9999      170      180   
15     150     150   -9999     160      150      150      180      230   
16     200     210     190   -9999      180      170    -9999      180   
17     220     230     250     240      210    -9999      230    -9999   
18     210     210     220   -9999      240      260    -9999      210   
19     250   -9999   -9999     250      260      260      240      230   
20     230   -9999     200     220    -9999      200    -9999      220   
21   -9999     230     230     230    -9999      230    -9999      210   
22   -9999

In [91]:
ghcnd.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2642 entries, 0 to 2641
Data columns (total 35 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   id       2642 non-null   object
 1   year     2642 non-null   int64 
 2   month    2642 non-null   int64 
 3   element  2642 non-null   object
 4   value1   2642 non-null   int64 
 5   value2   2642 non-null   int64 
 6   value3   2642 non-null   int64 
 7   value4   2642 non-null   int64 
 8   value5   2642 non-null   int64 
 9   value6   2642 non-null   int64 
 10  value7   2642 non-null   int64 
 11  value8   2642 non-null   int64 
 12  value9   2642 non-null   int64 
 13  value10  2642 non-null   int64 
 14  value11  2642 non-null   int64 
 15  value12  2642 non-null   int64 
 16  value13  2642 non-null   int64 
 17  value14  2642 non-null   int64 
 18  value15  2642 non-null   int64 
 19  value16  2642 non-null   int64 
 20  value17  2642 non-null   int64 
 21  value18  2642 non-null   int64 
 22  

In [108]:
global_min = 0
global_max = 0
for i in range(1, 32): # 32 is not included
    min = ghcnd[f'value{i}'].min()
    max = ghcnd[f'value{i}'].max()
    if min < global_min:
        global_min = min

    if max > global_max:
        global_max = max
        
print(f'global min: {global_min}, max: {global_max}')

global min: -9999, max: 422


We can assume that float16 is enough

In [115]:
# change some datatypes where possible
ghcnd['year'] = ghcnd['year'].astype('float16', errors='raise')
ghcnd['month'] = ghcnd['month'].astype('float16', errors='raise')
for i in range(1, 32): # 32 is not included
    ghcnd[f'value{i}'] = ghcnd[f'value{i}'].astype('float16', errors='raise')


ghcnd.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2642 entries, 0 to 2641
Data columns (total 35 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   id       2642 non-null   object 
 1   year     2642 non-null   float16
 2   month    2642 non-null   float16
 3   element  2642 non-null   object 
 4   value1   2642 non-null   float16
 5   value2   2642 non-null   float16
 6   value3   2642 non-null   float16
 7   value4   2642 non-null   float16
 8   value5   2642 non-null   float16
 9   value6   2642 non-null   float16
 10  value7   2642 non-null   float16
 11  value8   2642 non-null   float16
 12  value9   2642 non-null   float16
 13  value10  2642 non-null   float16
 14  value11  2642 non-null   float16
 15  value12  2642 non-null   float16
 16  value13  2642 non-null   float16
 17  value14  2642 non-null   float16
 18  value15  2642 non-null   float16
 19  value16  2642 non-null   float16
 20  value17  2642 non-null   float16
 21  value18  2642 

We can see the memory usage drop from ~723 KB to ~212 KB

### Determine a percentagem de null values em cada uma das variáveis.

In [124]:
# let's convert -9999 to null, due to datatype conversion, -9999 becomes -10000
for i in range(1, 32): # 32 is not included
    ghcnd[f'value{i}'] = ghcnd[f'value{i}'].replace(-10000.0, np.nan)
ghcnd

id    year  month element  value1  value2  value3  value4  \
0              HORTA  1973.0    1.0    TMAX     NaN   110.0     NaN   170.0   
1              HORTA  1973.0    2.0    TMAX   170.0   160.0   120.0   110.0   
2              HORTA  1973.0    3.0    TMAX     NaN   150.0   160.0     NaN   
3              HORTA  1973.0    4.0    TMAX   180.0   170.0     NaN   190.0   
4              HORTA  1973.0    5.0    TMAX   140.0     NaN     NaN   150.0   
...              ...     ...    ...     ...     ...     ...     ...     ...   
2637  CASTELO BRANCO  2018.0    8.0    TMAX     NaN   422.0     NaN     NaN   
2638  CASTELO BRANCO  2018.0    9.0    TMAX     NaN     NaN     NaN     NaN   
2639  CASTELO BRANCO  2018.0   10.0    TMAX   297.0   283.0     NaN     NaN   
2640  CASTELO BRANCO  2019.0    3.0    TMAX     NaN     NaN     NaN     NaN   
2641  CASTELO BRANCO  2019.0    7.0    TMAX     NaN     NaN     NaN     NaN   

      value5  value6  value7  value8  value9  value10  value11  value12  \
0      160.0   160.0     NaN     NaN     NaN    160.0    170.0    140.0   
1      130.0     NaN     NaN   160.0   150.0      NaN      NaN      NaN   
2      160.0     NaN   160.0   140.0   130.0    120.0      NaN    140.0   
3      170.0     NaN   190.0     NaN   160.0      NaN    180.0    180.0   
4        NaN   190.0   190.0     NaN     NaN    190.0      NaN    180.0   
...      ...     ...     ...     ...     ...      ...      ...      ...   
2637     NaN     NaN     NaN     NaN     NaN      NaN      NaN      NaN   
2638     NaN     NaN     NaN     NaN     NaN      NaN      NaN      NaN   
2639   311.0     NaN     NaN     NaN     NaN      NaN      NaN      NaN   
2640     NaN     NaN     NaN     NaN     NaN      NaN      NaN      NaN   
2641   290.0     NaN     NaN     NaN     NaN      NaN      NaN      NaN   

      value13  value14  value15  value16  value17  value18  value19  value20  \
0         NaN      NaN    180.0    140.0      NaN      NaN    150.0      NaN   
1       150.0    170.0      NaN    170.0    170.0    180.0      NaN    150.0   
2       150.0    160.0    170.0    150.0    160.0    160.0    160.0    190.0   
3         NaN    150.0      NaN    160.0      NaN    150.0    180.0    170.0   
4         NaN    180.0    190.0    190.0    170.0    190.0    170.0    160.0   
...       ...      ...      ...      ...      ...      ...      ...      ...   
2637      NaN      NaN      NaN      NaN      NaN      NaN      NaN      NaN   
2638      NaN      NaN    333.0    334.0      NaN      NaN      NaN      NaN   
2639      NaN      NaN      NaN      NaN    221.0      NaN    212.0      NaN   
2640    157.0      NaN      NaN      NaN      NaN      NaN      NaN      NaN   
2641      NaN      NaN      NaN      NaN      NaN      NaN      NaN      NaN   

      value21  value22  value23  value24  value25  value26  value27  value28  \
0       170.0      NaN      NaN      NaN    130.0      NaN      NaN      NaN   
1         NaN      NaN    150.0      NaN    170.0    180.0    170.0    170.0   
2       160.0      NaN      NaN    160.0    170.0    170.0    180.0      NaN   
3       170.0      NaN    160.0    170.0      NaN      NaN    170.0    170.0   
4         NaN    180.0      NaN    190.0      NaN    190.0    180.0    190.0   
...       ...      ...      ...      ...      ...      ...      ...      ...   
2637      NaN      NaN      NaN      NaN      NaN      NaN      NaN      NaN   
2638    327.0      NaN      NaN      NaN      NaN    308.0      NaN      NaN   
2639      NaN      NaN      NaN      NaN      NaN      NaN      NaN      NaN   
2640    177.0      NaN      NaN      NaN      NaN      NaN      NaN      NaN   
2641      NaN      NaN      NaN      NaN      NaN      NaN      NaN      NaN   

      value29  value30  value31  
0         NaN    180.0      NaN  
1         NaN      NaN      NaN  
2         NaN      NaN    180.0  
3       180.0    180.0      NaN  
4         NaN    210.0      NaN  
...       ...      ...      ...  
2637    

In [128]:
null_perc = ghcnd.isnull().mean() * 100

null_df = pd.DataFrame(null_perc)

null_df

0
id        0.000000
year      0.000000
month     0.000000
element   0.000000
value1   11.279334
value2   11.392884
value3   11.619985
value4   10.938683
value5   11.355034
value6   12.376987
value7   11.165783
value8   11.619985
value9   11.468584
value10  11.203634
value11  10.446631
value12  11.127933
value13  11.468584
value14  12.604088
value15  12.679788
value16  11.695685
value17  11.809235
value18  12.376987
value19  11.014383
value20  11.619985
value21  11.279334
value22  12.225587
value23  12.452687
value24  11.809235
value25  11.884936
value26  11.582135
value27  11.241484
value28  11.960636
value29  17.562453
value30  18.735806
value31  48.372445

### Crie uma nova coluna denominada count_nan que registe para cada observação o número de dias com temperaturas máximas nulas (NaN).

In [132]:
for i, row in ghcnd.iterrows():
    days_nan = 0
    for x in range(1, 32): # 32 is not included
        if pd.isna(ghcnd.loc[i, f'value{x}']):
            days_nan += 1
    ghcnd.loc[i, 'count_nan'] = days_nan

ghcnd

id    year  month element  value1  value2  value3  value4  \
0              HORTA  1973.0    1.0    TMAX     NaN   110.0     NaN   170.0   
1              HORTA  1973.0    2.0    TMAX   170.0   160.0   120.0   110.0   
2              HORTA  1973.0    3.0    TMAX     NaN   150.0   160.0     NaN   
3              HORTA  1973.0    4.0    TMAX   180.0   170.0     NaN   190.0   
4              HORTA  1973.0    5.0    TMAX   140.0     NaN     NaN   150.0   
...              ...     ...    ...     ...     ...     ...     ...     ...   
2637  CASTELO BRANCO  2018.0    8.0    TMAX     NaN   422.0     NaN     NaN   
2638  CASTELO BRANCO  2018.0    9.0    TMAX     NaN     NaN     NaN     NaN   
2639  CASTELO BRANCO  2018.0   10.0    TMAX   297.0   283.0     NaN     NaN   
2640  CASTELO BRANCO  2019.0    3.0    TMAX     NaN     NaN     NaN     NaN   
2641  CASTELO BRANCO  2019.0    7.0    TMAX     NaN     NaN     NaN     NaN   

      value5  value6  value7  value8  value9  value10  value11  value12  \
0      160.0   160.0     NaN     NaN     NaN    160.0    170.0    140.0   
1      130.0     NaN     NaN   160.0   150.0      NaN      NaN      NaN   
2      160.0     NaN   160.0   140.0   130.0    120.0      NaN    140.0   
3      170.0     NaN   190.0     NaN   160.0      NaN    180.0    180.0   
4        NaN   190.0   190.0     NaN     NaN    190.0      NaN    180.0   
...      ...     ...     ...     ...     ...      ...      ...      ...   
2637     NaN     NaN     NaN     NaN     NaN      NaN      NaN      NaN   
2638     NaN     NaN     NaN     NaN     NaN      NaN      NaN      NaN   
2639   311.0     NaN     NaN     NaN     NaN      NaN      NaN      NaN   
2640     NaN     NaN     NaN     NaN     NaN      NaN      NaN      NaN   
2641   290.0     NaN     NaN     NaN     NaN      NaN      NaN      NaN   

      value13  value14  value15  value16  value17  value18  value19  value20  \
0         NaN      NaN    180.0    140.0      NaN      NaN    150.0      NaN   
1       150.0    170.0      NaN    170.0    170.0    180.0      NaN    150.0   
2       150.0    160.0    170.0    150.0    160.0    160.0    160.0    190.0   
3         NaN    150.0      NaN    160.0      NaN    150.0    180.0    170.0   
4         NaN    180.0    190.0    190.0    170.0    190.0    170.0    160.0   
...       ...      ...      ...      ...      ...      ...      ...      ...   
2637      NaN      NaN      NaN      NaN      NaN      NaN      NaN      NaN   
2638      NaN      NaN    333.0    334.0      NaN      NaN      NaN      NaN   
2639      NaN      NaN      NaN      NaN    221.0      NaN    212.0      NaN   
2640    157.0      NaN      NaN      NaN      NaN      NaN      NaN      NaN   
2641      NaN      NaN      NaN      NaN      NaN      NaN      NaN      NaN   

      value21  value22  value23  value24  value25  value26  value27  value28  \
0       170.0      NaN      NaN      NaN    130.0      NaN      NaN      NaN   
1         NaN      NaN    150.0      NaN    170.0    180.0    170.0    170.0   
2       160.0      NaN      NaN    160.0    170.0    170.0    180.0      NaN   
3       170.0      NaN    160.0    170.0      NaN      NaN    170.0    170.0   
4         NaN    180.0      NaN    190.0      NaN    190.0    180.0    190.0   
...       ...      ...      ...      ...      ...      ...      ...      ...   
2637      NaN      NaN      NaN      NaN      NaN      NaN      NaN      NaN   
2638    327.0      NaN      NaN      NaN      NaN    308.0      NaN      NaN   
2639      NaN      NaN      NaN      NaN      NaN      NaN      NaN      NaN   
2640    177.0      NaN      NaN      NaN      NaN      NaN      NaN      NaN   
2641      NaN      NaN      NaN      NaN      NaN      NaN      NaN      NaN   

      value29  value30  value31  count_nan  
0         NaN    180.0      NaN       18.0  
1         NaN      NaN      NaN       13.0  
2         NaN      NaN    180.0        9.0  
3       180.0    180.0      NaN       11.0  
4         NaN    210.

### Crie um novo dataframe que considere apenas os registos com um valor total de NaN inferior a 7.

In [133]:
nan_lt_7_df = ghcnd.loc[ghcnd['count_nan'] < 7].copy()

nan_lt_7_df

id    year  month element  value1  value2  value3  value4  \
6              HORTA  1973.0    7.0    TMAX   210.0     NaN   230.0   220.0   
26             HORTA  1975.0    3.0    TMAX   160.0   150.0   150.0     NaN   
27             HORTA  1975.0    4.0    TMAX     NaN   300.0   180.0   170.0   
30             HORTA  1975.0    7.0    TMAX   220.0     NaN   230.0   240.0   
31             HORTA  1975.0    8.0    TMAX   240.0     NaN   230.0   250.0   
...              ...     ...    ...     ...     ...     ...     ...     ...   
2613  CASTELO BRANCO  2015.0    6.0    TMAX   318.0   324.0   349.0   349.0   
2614  CASTELO BRANCO  2015.0    7.0    TMAX   293.0   304.0   322.0   343.0   
2615  CASTELO BRANCO  2015.0    8.0    TMAX     NaN   355.0   341.0   338.0   
2625  CASTELO BRANCO  2016.0    6.0    TMAX     NaN   278.0   247.0   256.0   
2626  CASTELO BRANCO  2016.0    7.0    TMAX   323.0   350.0   376.0   349.0   

      value5  value6  value7  value8  value9  value10  value11  value12  \
6      240.0   250.0   240.0   250.0   210.0    240.0    220.0    210.0   
26     170.0   180.0   170.0   150.0   180.0      NaN    180.0    180.0   
27       NaN   170.0   170.0   160.0   160.0    270.0    170.0    160.0   
30     250.0   250.0   250.0   250.0   250.0      NaN    240.0    250.0   
31     250.0   240.0   260.0   260.0   260.0    260.0    260.0    260.0   
...      ...     ...     ...     ...     ...      ...      ...      ...   
2613   337.0   361.0   331.0   343.0   323.0    300.0      NaN    211.0   
2614   340.0   368.0   379.0   376.0   370.0    331.0    354.0    355.0   
2615   348.0   358.0   335.0   349.0   374.0    377.0    326.0    289.0   
2625   265.0   275.0     NaN   339.0   320.0    262.0    272.0    317.0   
2626   357.0     NaN     NaN   356.0   379.0    372.0    327.0    308.0   

      value13  value14  value15  value16  value17  value18  value19  value20  \
6       220.0    240.0      NaN    220.0    210.0    210.0      NaN    240.0   
26      160.0    180.0    170.0    130.0    150.0    170.0    180.0    170.0   
27      170.0    180.0    180.0    180.0    170.0    160.0    160.0    180.0   
30      240.0    240.0    240.0    250.0    250.0    240.0    250.0    250.0   
31      240.0    250.0    260.0    270.0    260.0    270.0    240.0      NaN   
...       ...      ...      ...      ...      ...      ...      ...      ...   
2613    197.0    199.0    231.0    282.0    310.0    318.0    353.0    361.0   
2614    369.0    384.0    391.0    375.0    359.0    334.0    340.0    356.0   
2615    246.0    280.0      NaN    304.0    312.0    330.0    349.0      NaN   
2625    264.0    253.0    199.0    207.0    236.0    258.0    301.0      NaN   
2626    317.0    338.0      NaN      NaN    387.0    370.0    368.0    352.0   

      value21  value22  value23  value24  value25  value26  value27  value28  \
6       220.0    250.0      NaN    250.0      NaN    270.0    240.0    330.0   
26        NaN    170.0    170.0      NaN      NaN    180.0    150.0    160.0   
27      180.0    180.0    190.0    190.0    180.0    180.0    170.0    180.0   
30      250.0    250.0    260.0    250.0    250.0    260.0    260.0    260.0   
31      250.0    250.0      NaN    250.0    250.0    250.0    260.0    220.0   
...       ...      ...      ...      ...      ...      ...      ...      ...   
2613    360.0    303.0      NaN    308.0    341.0    360.0    390.0    399.0   
2614    352.0    369.0    376.0    298.0    349.0    368.0    367.0    373.0   
2615    350.0    308.0    247.0    249.0    311.0    303.0    301.0      NaN   
2625    360.0    344.0    303.0    311.0    322.0    319.0    345.0    366.0   
2626    330.0    322.0    347.0    367.0    390.0    395.0    375.0    387.0   

      value29  value30  value31  count_nan  
6       250.0    240.0    220.0        5.0  
26      170.0    170.0    150.0        5.0  
27      180.0    170.0      NaN        3.0  
30      260.0    250.0    240.0        2.0  
31      220.0    240.

### Reccorra agora a chunking para determinar a percentagem de null values em cada uma das variáveis na totalidade dos registos do dataframe.

In [169]:
run = False # change to True to run

if run:
    # Dividir em ficheiros mais pequenos, neste caso, com 300000 linhas
    chunk_size = 300000

    global_null_counts = None
    row_count = 0
    batch = 1
    for df_chunk in pd.read_csv(ghcnd_daily_csv_path, chunksize=chunk_size, low_memory=False):
        # calculate the null values for each column
        null_counts = pd.DataFrame(df_chunk.isnull().sum())
        if global_null_counts is None:
            global_null_counts = null_counts
        else:
            global_null_counts += null_counts

        row_count += len(df_chunk)
        
        print(f'processed batch # {batch}')
        batch += 1
    
    # divide values by total values to get percentages
    global_null_counts[0] = global_null_counts[0] / row_count

global_null_counts

processed batch # 1
processed batch # 2


0
id       0.000000
year     0.000000
month    0.000000
element  0.000000
value1   0.000000
...           ...
sflag30  0.166912
value31  0.000000
mflag31  1.000000
qflag31  0.999645
sflag31  0.472442

[128 rows x 1 columns]

### Determine a temperatura média para cada observação (estação/ano/mês) do dataframe. Guarde os valores obtidos numa nova coluna denominada daily_avg_temp. Não se esqueça de considerar apenas as colunas cujo nome começa por value.

In [194]:
run = False # change to True to run

if run:
    def mean_without_neg9999(row):
        filtered_row = row[row != -9999]  # filter out -9999 values
        if len(filtered_row) == 0:  # handle case where all values are -9999
            return np.nan
        else:
            return filtered_row.mean()
        
    daily_avg_tmp_csv_path = os.path.join(ignored_dir, 'daily_avg_temp.csv')
    if os.path.exists(daily_avg_tmp_csv_path):
        os.remove(daily_avg_tmp_csv_path) # clear csv
        
    # Dividir em ficheiros mais pequenos, neste caso, com 300000 linhas
    chunk_size = 300000

    batch = 1
    for df_chunk in pd.read_csv(ghcnd_daily_csv_path, chunksize=chunk_size, low_memory=False):
        df_chunk['daily_avg_temp'] = df_chunk.apply(lambda row: mean_without_neg9999(row.filter(like='value')), axis=1)

        df_chunk.to_csv(daily_avg_tmp_csv_path, mode='a', header=(True if batch == 1 else False), index=False)
        
        print(f'processed batch # {batch}')
        batch += 1

df_chunk

processed batch # 1
processed batch # 2


id  year  month element  value1  mflag1 qflag1 sflag1  value2  \
3000  AG000060390  1976      8    TMAX     301     NaN    NaN      E     280   
3001  AG000060390  1976      9    TMAX     305     NaN    NaN      E     300   
3002  AG000060390  1976     10    TMAX     257     NaN    NaN      E     230   
3003  AG000060390  1976     11    TMAX     197     NaN    NaN      E     195   
3004  AG000060390  1976     12    TMAX     180     NaN    NaN      E     229   
...           ...   ...    ...     ...     ...     ...    ...    ...     ...   
5995  AG000060680  1996      2    TMAX     250     NaN    NaN      G     266   
5996  AG000060680  1996      3    TMAX     223     NaN    NaN      G     218   
5997  AG000060680  1996      4    TMAX     302     NaN    NaN      G     302   
5998  AG000060680  1996      5    TMAX     302     NaN    NaN      G     302   
5999  AG000060680  1996      6    TMAX     322     NaN    NaN      G     326   

      mflag2 qflag2 sflag2  value3  mflag3 qflag3 sflag3  value4  mflag4  \
3000     NaN    NaN      E     272     NaN    NaN      E     286     NaN   
3001     NaN    NaN      E     306     NaN    NaN      E     269     NaN   
3002     NaN    NaN      E     250     NaN    NaN      E     260     NaN   
3003     NaN    NaN      E     204     NaN    NaN      E     161     NaN   
3004     NaN    NaN      E     226     NaN    NaN      E     155     NaN   
...      ...    ...    ...     ...     ...    ...    ...     ...     ...   
5995     NaN    NaN      G     266     NaN    NaN      G     269     NaN   
5996     NaN    NaN      G     233     NaN    NaN      G     255     NaN   
5997     NaN    NaN      G     305     NaN    NaN      G     304     NaN   
5998     NaN    NaN      G     325     NaN    NaN      G     322     NaN   
5999     NaN    NaN      G     304     NaN    NaN      G     293     NaN   

     qflag4 sflag4  value5  mflag5 qflag5 sflag5  value6  mflag6 qflag6  \
3000    NaN      E     363     NaN    NaN      E     372     NaN    NaN   
3001    NaN      E     279     NaN    NaN      E     250     NaN    NaN   
3002    NaN      E     267     NaN    NaN      E     259     NaN    NaN   
3003    NaN      E     190     NaN    NaN      E     217     NaN    NaN   
3004      I      E     153     NaN    NaN      E     210     NaN    NaN   
...     ...    ...     ...     ...    ...    ...     ...     ...    ...   
5995    NaN      G     257     NaN    NaN      G     232     NaN    NaN   
5996    NaN      G     267     NaN    NaN      G     240     NaN    NaN   
5997    NaN      G     306     NaN    NaN      G     287     NaN    NaN   
5998    NaN      G     335     NaN    NaN      G     358     NaN    NaN   
5999    NaN      G     274     NaN    NaN      G     256     NaN    NaN   

     sflag6  value7  mflag7 qflag7 sflag7  value8  mflag8 qflag8 sflag8  \
3000      E     327     NaN    NaN      E     350     NaN    NaN      E   
3001      E     261     NaN    NaN      E     265     NaN    NaN      E   
3002      E     263     NaN    NaN      E     257     NaN    NaN      E   
3003      E     197     NaN    NaN      E     198     NaN    NaN      E   
3004      E     219     NaN    NaN      E     233     NaN    NaN      E   
...     ...     ...     ...    ...    ...     ...     ...    ...    ...   
5995      G     258     NaN    NaN      G     245     NaN    NaN      G   
5996      G     272     NaN    NaN      G     274     NaN    NaN      G   
5997      G     290     NaN    NaN      G     290     NaN    NaN      G   
5998      G     359     NaN    NaN      G     362     NaN    NaN      G   
5999      G     335     NaN    NaN      G     342     NaN    NaN      G   

      value9  mflag9 qflag9 sflag9  value10  mflag10 qflag10 sflag10  value11  \
3000     345     NaN    NaN      E      300      NaN     NaN       E      295   
3001     297     NaN    NaN      E      286      NaN     NaN       E      295   
3002     277     NaN    NaN      E      280      NaN     NaN       E      263   
3003     261   

### Agrupe os dados por nome de estação e ano aplicando a função mean. Mostre apenas os resultados para a coluna daily_avg_temp.

In [195]:
df_mean = df_chunk.groupby(['id', 'year'])['daily_avg_temp'].mean()

df_mean

id           year
AG000060390  1976    238.323441
             1977    232.425544
             1978    228.689747
             1979    228.120411
             1980    227.181056
                        ...    
AG000060680  1992    280.266080
             1993    286.013652
             1994    284.855690
             1995    287.832124
             1996    281.376523
Name: daily_avg_temp, Length: 264, dtype: float64